File: Blind75/leetcode_133.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

from collections import deque
from typing import Optional

class Solution:
    nodeMap = dict()
    visitedMap = dict()

    def printGraph(self, node: Optional['Node']) -> None:
        if node:
            self.visitedMap.clear()
            q = deque()
            q.append(node)

            # print(f"neighbors: {[n.val if n else None for n in node.neighbors]}")
            while len(q) > 0:
                curr = q.popleft()

                print(f"Node: {curr.val}")
                print(f"Neighbors: ")
                for n in curr.neighbors:
                    print(f"{n.val} ")
                    if self.visitedMap.get(n.val) is None:
                        q.append(n)
                
                self.visitedMap[curr.val] = True
            
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:

        def deepClone(node: Optional['Node']) -> Optional['Node']:
            q = deque()
            q.append(node)
            _node = Node(node.val)
            self.nodeMap[node.val] = _node
            newGraph = _node

            while len(q) > 0:
                currNode = q.popleft()
                print(f"current node val: {currNode.val}")
                _currNode = self.nodeMap[currNode.val]

                if self.visitedMap.get(currNode.val) is None:
                    for neighbor in currNode.neighbors:
                        print(f"next neighbor to be queued: {neighbor.val}")
                        q.append(neighbor)

                        _neighbor = None
                        if self.nodeMap.get(neighbor.val) is None:
                            self.nodeMap[neighbor.val] = Node(neighbor.val)
                        
                        _neighbor = self.nodeMap[neighbor.val]
                        _currNode.neighbors.append(_neighbor)
                        
                    
                    self.visitedMap[currNode.val] = True
            
            return newGraph
        
        if node is None:
            return None
        if node.neighbors == []:
            return Node(node.val)
        graph = deepClone(node)
        self.printGraph(graph)
        return graph

soln = Solution()
node1 = Node(1)
node2 = Node(2)
node3 = Node(3)
node4 = Node(4)

# node1.neighbors = [node2, node4]
# node2.neighbors = [node1, node3]
# node3.neighbors = [node2, node4]
# node4.neighbors = [node1, node3]

node1.neighbors = [node2]
node2.neighbors = [node1]

soln.cloneGraph(node1)

File: Blind75/leetcode_133.cpp
#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <set>

class Node
{
public:
    int val;
    std::vector<Node *> neighbors;

    Node()
    {
        this->val = 0;
        this->neighbors = std::vector<Node *>();
    }

    Node(int val)
    {
        this->val = val;
        this->neighbors = std::vector<Node *>();
    }

    Node(int val, std::vector<Node *> neighbors)
    {
        this->val = val;
        this->neighbors = neighbors;
    }
};

class Solution
{
public:
    void printGraph(Node *node)
    {
        std::set<int> visited;
        std::queue<Node *> q;
        q.push(node);

        while (q.size() > 0)
        {
            Node *curr = q.front();
            q.pop();

            std::cout << "curr node:" << curr->val << std::endl;
            if (visited.count(curr->val) == 0)
            {
                std::vector<Node *> neighbors = curr->neighbors;
                std::cout << "neighbors: ";
                // for (auto it = neighbors.begin(); it != neighbors.end(); ++it)
                // {
                //     std::cout << it->val << " ";
                //     std::cout << *it << "";
                //     zsq.push(it);
                // }
                for (Node *neighbor : neighbors)
                {
                    std::cout << neighbor->val << " ";
                    // std::cout <<  << "";
                    q.push(neighbor);
                }
                std::cout << std::endl;
                visited.insert(curr->val);
            }
        }
    }

    Node *clone(Node *node)
    {
        std::map<int, Node *> nodeMap;
        std::set<int> visited;
        std::queue<Node *> q;
        q.push(node);

        Node *_node = new Node(node->val);
        nodeMap[node->val] = _node;

        while (q.size() > 0)
        {
            Node *currNode = q.front();
            q.pop();
            Node *_currNode = nodeMap[currNode->val];

            if (visited.count(currNode->val) == 0)
            {
                std::vector<Node *> neighbors = currNode->neighbors;
                for (Node *neighbor : neighbors)
                {
                    q.push(neighbor);
                    if (nodeMap.count(neighbor->val) == 0)
                    {
                        nodeMap[neighbor->val] = new Node(neighbor->val);
                    }
                    _currNode->neighbors.push_back(nodeMap[neighbor->val]);
                }

                visited.insert(currNode->val);
            }
        }

        return _node;
    }

    Node *cloneGraph(Node *node)
    {
        if (node == nullptr)
        {
            return nullptr;
        }

        if (node->neighbors.size() == 0)
        {
            return new Node(node->val);
        }

        Node *cloned = clone(node);
        printGraph(cloned);
        return cloned;
        // return clone(node);
    }
};

int main()
{
    Node *node1 = new Node(1);
    Node *node2 = new Node(2);
    Node *node3 = new Node(3);
    Node *node4 = new Node(4);

    node1->neighbors.push_back(node2);
    node1->neighbors.push_back(node4);

    node2->neighbors.push_back(node1);
    node2->neighbors.push_back(node3);

    node3->neighbors.push_back(node2);
    node3->neighbors.push_back(node4);

    node4->neighbors.push_back(node1);
    node4->neighbors.push_back(node3);

    Solution soln;
    soln.cloneGraph(node1);
    return 0;
}